src/c-writer.cc

```diff
@@ -1537,20 +1584,30 @@ void CWriter::WriteElemInitializers() {
   Write(Newline(), "static void init_tables(", ModuleInstanceTypeName(),
         "* instance) ", OpenBrace());
 
-  const Table* table = module_->tables.empty() ? nullptr : module_->tables[0];
-
-  if (table && module_->num_table_imports == 0) {
-    uint32_t max =
-        table->elem_limits.has_max ? table->elem_limits.max : UINT32_MAX;
-    Write("wasm_rt_allocate_table(", ExternalInstancePtr(table->name), ", ",
-          table->elem_limits.initial, ", ", max, ");", Newline());
+  if (module_->tables.size() > module_->num_table_imports) {
+    Index table_idx = module_->num_table_imports;
+    for (Index i = table_idx; i < module_->tables.size(); i++) {
+      const Table* table = module_->tables[i];
+      uint32_t max =
+          table->elem_limits.has_max ? table->elem_limits.max : UINT32_MAX;
+      Write("wasm_rt_allocate_", GetReferenceTypeName(table->elem_type),
+            "_table(", ExternalInstancePtr(table->name), ", ",
+            table->elem_limits.initial, ", ", max, ");", Newline());
+    }
   }
+
   for (const ElemSegment* elem_segment : module_->elem_segments) {
     if (elem_segment->kind != SegmentKind::Active) {
       continue;
     }
 
-    Write("table_init(", ExternalInstancePtr(table->name), ", ");
+    const Table* table = module_->GetTable(elem_segment->table_var);
+
+    if (table->elem_type != Type::FuncRef) {
+      WABT_UNREACHABLE;
```

Comment:
Don't we have tests for elem segments that contain extern refs?

Reply:
No... I don't think it's (currently) possible to represent an elem segment of externrefs. I think they are generally opaque to the Wasm module; only the host knows how to create one.

Reply:
Presumably a module could declare an element segment containing null externrefs?  Not needed for PR either way... since all tests pass.

Reply:
This is... hmm. This is not supposed to be possible (https://webassembly.github.io/spec/core/syntax/modules.html#element-segments: "In the current version of WebAssembly, only tables of element type can be initialized with an element segment. This limitation may be lifted in the future."), and in the binary format, there is currently no way to represent an element segment of any type other than funcref (https://webassembly.github.io/spec/core/binary/modules.html#element-section defines elemkind 0x00 as funcref and no other elemkinds are defined).

However, I can't find where this same formal limitation is imposed on the text format. wat2wasm currently accepts an active element segment of type externref, but then wasm-validate rejects the resulting module. I think this needs further study -- I will look into it.

Reply:
Submitted https://github.com/WebAssembly/spec/issues/1543